Quantum channels

chriseclectic opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->
Closes #1193 

### Summary

Adds quantum channel package to `quantum_info`. This package contains classes for representing quantum channels (CPTP maps, and also general Non-CPTP operator maps) and is essentially an implementation of the paper: 

*Tensor networks and graphical calculus for open quantum systems*
Wood, Biamonte, Cory, Quant. Inf. Comp. 15, 0579-0811 (2015)
Open access: [arXiv: arXiv:1111.6950v3](https://arxiv.org/abs/1111.6950)

### Details and comments

* Contains classes for channel representations: 
  * `Choi`: Choi-Jamiolkowski process matrix
  * `SuperOp`: Linear superoperator
  * `Chi`: Pauli-basis Chi matrix
  * `PTM`: Pauli transfer-matrix superoperator
  * `Kraus`:  Kraus / operator-sum operators
  * `Stinespring`: Stinespring / system-environment insometry
  * `UnitaryChannel`: single-matrix Kraus representation (not generic operator map like other reps, and not exposed to user by default)

* The classes can be converted into each other using their init methods eg: `SuperOp(Kraus([k1, k2]))`
* All classes contain methods for
  * `is_cptp`: check if map is completely positive trace preserving
  * `evolve`: evolve a quantum state by the channel
  * `adjoint`: compute to the adjoint channel
  * `conjugate`: compute the conjugate channel
  * `transpose`: compute the transposed channel
  * `compose`: compose with another channel B(A(rho))
  * `tensor`: tensor product with another channel A \otimes B
  * `expand`: expand with another channel B \otimes A (reverse order tensor product)
  * `power`: compose channel with itself n times
  * `add`: linear addition of two channels
  * `subtract`: linear subtraction of two channels
  * `multiply`: scalar multiplication of a channel
* Operator overloads:
  * `+,-,*` for add, subtract, multiply
  * `@` for compose
  * `^` for tensor

Closes #1193 

### Tests

- [x] Tests for `Choi` class
- [x] Tests for `SuperOp` class
- [x] Tests for `Chi` class
- [x] Tests for `PTM` class
- [x] Tests for `Stinespring` class
- [x] Tests for `Kraus` class
- [x] Tests for `UnitaryChannel` helper class
- [x] Equivalence tests for evolution
- [x] Equivalence tests for tensor, expand
- [x] Equivalence tests for compose
- [x] Equivalence tests for conjugate, transpose, adjoint
- [x] Equivalence tests for add, subtract, multiply

### Future PR:

- Integrate with Ericks `Unitary` class in PR #1961 
- Integrate with Qiskit Aer noise model

ajavadia commented:
I think this closes #1193

chriseclectic commented:
@jaygambetta @ajavadia this is finished and ready for review

jaygambetta commented on test/python/quantum_info/channel/test_chi.py:
```diff
@@ -0,0 +1,433 @@
+        # UnitaryChannel evolution
+        chan1 = Chi(self.chiX)
+        chan2 = Chi(self.chiY)
+        chan = chan1.compose(chan2)
```
I not sure I really like this order

chriseclectic replied:
We need this order to be consistent with the rest of the circuit gate type objects.

jaygambetta commented:
also chris can we make an issue to make the PTM change order depending on the order a user gives it. Ie I should be able to put in a Pauli group.

chriseclectic commented:
Changed order of tensor product and removed the `front` kwarg to swap order. The reverse order tensor product is now done with a seperate method called `expand`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=1915) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=1915) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=1915) it.</sub>

ajavadia commented on test/python/quantum_info/channel/base.py:
```diff
@@ -0,0 +1,120 @@
+                       atol=1e-6,
+                       equal_nan=False,
+                       msg=None):
+        """Assert to objects are equal using Numpy.allclose."""
```
```suggestion
        """Assert two objects are equal using Numpy.allclose."""
```

ajavadia commented:
isn't there a bunch of duplication between the UnitaryChannel here and the one that derives from Gate in #1774 ?

chriseclectic commented:
@ajavadia yeah it  can be removed once Ericks is finished, but its needed here internally for tests.

ajavadia commented on qiskit/quantum_info/operators/predicates.py:
```diff
@@ -0,0 +1,158 @@
+    return np.allclose(mat, np.conj(mat.T), rtol=rtol, atol=atol)
+
+
+def is_positive_semidefinite_matrix(op, rtol=RTOL_DEFAULT, atol=ATOL_DEFAULT):
```
nit: can you make the arg of all these predicate functions `mat`? currently some are `op` and some are `mat`.

ajavadia commented:
Nice work, especially writing all those unit tests.

This generally looks good to me. I have one concern however: what is the actual use case of the `inplace` args? At first I liked the idea, but as I was working on #1816 I realized this is not as easy to do, since:
- we don't want to allow modifying the fundamental objects (say H gate). 
- it becomes a bit ambiguous in hierarchical setting (modifying circuit A in place, when it contains circuit B; does B get modified?). 
- it complicates the code a lot, as you have `if inplace:` checks all over.
- we open up circuit/channel objects to being modified, which may potentially cause hard-to-track bugs later.

So what do we lose if we just said every operation returns a new circuit or channel? I'm mainly talking about the end user scenario, not internal implementation (cc'ing @kdk).
